substra/sdk/backends/local/dal.py

```diff
@@ -0,0 +1,137 @@
+
+from substra.sdk import exceptions, schemas
+from substra.sdk.backends.remote import backend
+from substra.sdk.backends.local import db, models
+
+
+class DataAccess:
+    """Data access layer.
+
+    This is an intermediate layer between the backend and the local/remote data access.
+    """
+
+    def __init__(self, remote_backend: typing.Optional[backend.Remote]):
+        self._db = db.get()
+        self._remote = remote_backend
+        self._tmp_dir = tempfile.TemporaryDirectory(prefix="/tmp/")
+
+    @property
+    def tmp_dir(self):
+        return pathlib.Path(self._tmp_dir.name)
+
+    @staticmethod
+    def _is_local(key: str):
+        """Check if the key corresponds to a local
+        or remote asset.
+        """
+        return key.startswith("local_")
+
+    def _get_asset_name(self, type_):
+        if type_ in [
+            schemas.Type.Algo,
+            schemas.Type.AggregateAlgo,
+            schemas.Type.CompositeAlgo
+        ]:
+            asset_name = "algo.tar.gz"
+            field_name = "content"
+
+        elif type_ == schemas.Type.Dataset:
+            asset_name = "opener.py"
+            field_name = "opener"
+
+        elif type_ == schemas.Type.Objective:
+            asset_name = "metrics.zip"
+            field_name = "metrics"
+
+        else:
+            raise ValueError(f"Cannot download this type of asset {type_}")
+
+        return asset_name, field_name
+
+    def _get_response(self, type_, asset):
+        return models.SCHEMA_TO_MODEL[type_].from_response(asset)
+
+    def login(self, username, password):
+        if self._remote:
+            self._remote.login(username, password)
+
+    def add(self, asset, exist_ok: bool = False):
+        return self._db.add(asset, exist_ok=exist_ok)
+
+    def remote_download(self, asset_type, url_field_path, key, destination):
+        self._remote.download(asset_type, url_field_path, key, destination)
+
+    def get_remote_description(self, asset_type, key):
+        return self._remote.describe(asset_type, key)
+
+    def get_with_files(self, type_: schemas.Type, key: str):
+        """Get the asset with files on the local disk for execution"""
+        if self._is_local(key):
+            return self._db.get(type_, key)
+        else:
+            asset_name, field_name = self._get_asset_name(type_)
+            asset = self._get_response(
+                type_,
+                self._remote.get(type_, key)
+            )
+            tmp_directory = self.tmp_dir / key
+            description_path = tmp_directory / "description.md"
+            asset_path = tmp_directory / asset_name
+
+            if not tmp_directory.exists():
+                pathlib.Path.mkdir(tmp_directory)
+
+                self._remote.download(
+                    type_,
+                    field_name + ".storageAddress",
+                    key,
+                    asset_path,
+                )
+
+                description = self._remote.describe(type_, key)
```

Comment:
We should ignore the asset description as it's not needed for its execution.

Reply:
One less call to the remote backend and a faster execution :)